game-core/src/main/java/games/strategy/triplea/settings/ClientSetting.java

```diff
@@ -211,13 +205,26 @@ public static void flush() {
     }
   }
 
+  private static Preferences getPreferences() {
+    return Optional.ofNullable(preferencesRef.get())
+        .orElseThrow(() -> new IllegalStateException("ClientSetting framework has not been initialized. "
+            + "Did you forget to call ClientSetting#initialize() in production code "
+            + "or ClientSetting#setPreferences() in test code?"));
+  }
+
+  @VisibleForTesting
+  static void setPreferences(final Preferences preferences) {
+    preferencesRef.set(preferences);
+  }
+
   @Override
   public boolean isSet() {
     return !value().trim().isEmpty();
   }
 
   @Override
   public void save(final String newValue) {
+    onSaveActions.forEach(saveAction -> saveAction.accept(newValue));
```

Comment:
Also a location where a `checkNotNull()` precondition may be appropriate, just so a `null` doesn't escape into the listeners.  Although, it will still "almost" fail-fast because the call to `Preferences#put()` on the next line will throw an NPE if `newValue` is `null`.

Reply:
Filechooser has an example where an empty value is valid input. Also maybe for clearing out a value. We could change API to: `Consumer<Optional<String>>`

But, the original ClientSettings API used String and made decision that `Optional.empty` was equivalent to both `null` and the empty string. 

We can convert nulls to empty string here to help guarantee this. Ideally, it would be enforced upstream.

Reply:
null to empty string guard added: 07347f9b7

Reply:
Just to make sure I understand your intention here...  You want to allow `null` values and just silently convert them to empty strings?  If so, line 228 will still trigger an NPE if `newValue` is `null`:

```java
getPreferences().put(name(), newValue);
```

Reply:
That would be a problem, I updated the code to do a 'remove' if `newValue` is null: cda8f428a

The original contract of `ClientSettings` is that saved values are always non-null. FileChooser I think broke that potentially, but it might also never return null. It may be the case that we never see null passed to this method. If it is not, I do not want to refactor that now to fix it. So to maintain that original contract, that `null` and `empty string` are equivalent, they are both empty option preference, we can convert null to empty string here to at least guarantee it from this point on.

The alternative is to convert to a `Consumer<Optional<String>>`, but original clientSettings already had that consideration, I recall I chose to go for empty string to be treated the same as an Optional.empty, so no Optionals are required.